src/backend/distributed/utils/colocation_utils.c

```diff
@@ -14,14 +14,332 @@
+	}
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * MarkTablesColocated puts both tables to same colocation group. If the
+ * source table is in INVALID_COLOCATION_ID group, then it creates a new
+ * colocation group and assigns both tables to same colocation group. Otherwise,
+ * it adds the target table to colocation group of the source table.
+ */
+static void
+MarkTablesColocated(Oid sourceRelationId, Oid targetRelationId)
+{
+	uint32 sourceColocationId = INVALID_COLOCATION_ID;
+	Relation pgDistColocation = NULL;
+	Var *sourceDistributionColumn = NULL;
+	Var *targetDistributionColumn = NULL;
+	Oid sourceDistributionColumnType = InvalidOid;
+	Oid targetDistributionColumnType = InvalidOid;
+
+	CheckHashPartitionedTable(sourceRelationId);
+	CheckHashPartitionedTable(targetRelationId);
+
+	sourceDistributionColumn = PartitionKey(sourceRelationId);
+	sourceDistributionColumnType = sourceDistributionColumn->vartype;
+
+	targetDistributionColumn = PartitionKey(targetRelationId);
+	targetDistributionColumnType = targetDistributionColumn->vartype;
+
+	if (sourceDistributionColumnType != targetDistributionColumnType)
+	{
+		char *sourceRelationName = get_rel_name(sourceRelationId);
+		char *targetRelationName = get_rel_name(targetRelationId);
+
+		ereport(ERROR, (errmsg("cannot colocate tables %s and %s",
+							   sourceRelationName, targetRelationName),
+						errdetail("Distribution column types don't match for "
+								  "%s and %s.", sourceRelationName,
+								  targetRelationName)));
+	}
+
+	/*
+	 * Get an exclusive lock on the colocation system catalog. Therefore, we
+	 * can be sure that there will no modifications on the colocation table
+	 * until this transaction is committed.
+	 */
+	pgDistColocation = heap_open(DistColocationRelationId(), ExclusiveLock);
+
+	/* check if shard placements are colocated */
+	ErrorIfShardPlacementsNotColocated(sourceRelationId, targetRelationId);
+
+	/*
+	 * Get colocation group of the source table, if the source table does not
+	 * have a colocation group, create a new one, and set it for the source table.
+	 */
+	sourceColocationId = TableColocationId(sourceRelationId);
+	if (sourceColocationId == INVALID_COLOCATION_ID)
+	{
+		uint32 shardCount = ShardIntervalCount(sourceRelationId);
+		uint32 shardReplicationFactor = TableShardReplicationFactor(sourceRelationId);
+
+		sourceColocationId = CreateColocationGroup(shardCount, shardReplicationFactor,
+												   sourceDistributionColumnType);
+		UpdateRelationColocationGroup(sourceRelationId, sourceColocationId);
+	}
+
+	/* finally set colocation group for the target relation */
+	UpdateRelationColocationGroup(targetRelationId, sourceColocationId);
+
+	heap_close(pgDistColocation, NoLock);
+}
+
+
+/*
+ * ErrorIfShardPlacementsNotColocated checks if the shard placements of the
+ * given two relations are physically colocated. If they are not colocated,
+ * it errors out.
+ *
+ * Note that, this functions assumes that both tables are hash distributed.
```

Comment:
Let's mention that we error out if the placements doesn't have the same health states?